core/src/main/java/org/springframework/security/access/expression/method/ExpressionBasedAuthorizationManagerAfterAdvice.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright 2002-2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.security.access.expression.method;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Method;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Supplier;
+
+import org.aopalliance.intercept.MethodInvocation;
+
+import org.springframework.aop.MethodMatcher;
+import org.springframework.aop.support.AopUtils;
+import org.springframework.aop.support.StaticMethodMatcher;
+import org.springframework.core.annotation.AnnotationUtils;
+import org.springframework.expression.EvaluationContext;
+import org.springframework.expression.Expression;
+import org.springframework.expression.ExpressionParser;
+import org.springframework.security.access.AccessDeniedException;
+import org.springframework.security.access.expression.ExpressionUtils;
+import org.springframework.security.access.method.AuthorizationManagerAfterAdvice;
+import org.springframework.security.access.method.MethodAuthorizationContext;
+import org.springframework.security.access.prepost.PostAuthorize;
+import org.springframework.security.access.prepost.PostFilter;
+import org.springframework.security.core.Authentication;
+import org.springframework.util.Assert;
+import org.springframework.util.StringUtils;
+
+/**
+ * An {@link AuthorizationManagerAfterAdvice} which can determine if an
+ * {@link Authentication} has access to the returned object from the
+ * {@link MethodInvocation} by evaluating expressions from the {@link PostFilter} and the
+ * {@link PostAuthorize} annotations.
+ *
+ * @author Evgeniy Cheban
+ */
+public final class ExpressionBasedAuthorizationManagerAfterAdvice
+		implements AuthorizationManagerAfterAdvice<MethodAuthorizationContext> {
+
+	private final MethodMatcher methodMatcher = new StaticMethodMatcher() {
+		@Override
+		public boolean matches(Method method, Class<?> targetClass) {
+			return resolveAttribute(method, targetClass) != null;
+		}
+	};
+
+	private final Map<Method, AuthorizationAttribute> cachedAttributes = new ConcurrentHashMap<>();
+
+	private final MethodSecurityExpressionHandler expressionHandler;
+
+	/**
+	 * Creates an instance.
+	 * @param expressionHandler the {@link MethodSecurityExpressionHandler} to use
+	 */
+	public ExpressionBasedAuthorizationManagerAfterAdvice(MethodSecurityExpressionHandler expressionHandler) {
```

Comment:
I'd recommend that this change to a setter. Typically, Spring Security places required parameters in the constructor and optional ones as setters. Since there's a default handler that will work in most cases, using the default constructor will make it more convenient for applications to use.

Reply:
Done.